### ISSUE TITLE
Flow forces instance reconstruction or does not type check

### ISSUE THREAD

**Issue by Gozala** [facebook/flow#1309]
Here is a the the most minimal test case as I was able to produce:

### `result.js.flow`

```js
/* @flow */

interface ResultInterface <x, a>
  { toString():string
  , capture <y> (handle:(error:x) => Result<y, a>):Result<y, a>
  , and <b> (result:Result<x, b>):Result<x, b>
  }

type Ok <value>
  = ResultInterface<*, value>
  & { type: "Ok"
    , value: value
    }

type Error <error>
  = ResultInterface<error, *>
  & { type: "Error"
    , error: error
    }

export type Result <error, value>
  = Error <error>
  | Ok <value>

declare export function ok <x, a>
  (value:a) :
  Result<x, a>

declare export function error <x, a>
  (error: x) :
  Result<x, a>
```

### `result.js`

```js
/* @flow */

/*::
import type {Result} from "./result"
export type {Result}
*/

class Ok <x, a> {
  /*::
  type: "Ok";
  value: a;
  */
  constructor(value:a) {
    this.value = value
  }
  capture /*::<y>*/ (handle/*:(error:x) => Result<y, a>*/)/*:Result<y,a>*/ {
    return this
  }
  and /*::<b>*/ (other/*:Result<x, b>*/)/*:Result<x, b>*/ {
     return other
  }
}
Ok.prototype.type = "Ok"

export const ok = /*::<x,a>*/
  (value/*:a*/)/*:Result<x,a>*/ =>
  new Ok(value)

class Error <x, a> {
  /*::
  type: "Error";
  error: x;
  */
  constructor(error:x) {
    this.error = error
  }
  capture /*::<y>*/ (handle/*:(error:x) => Result<y, a>*/)/*:Result<y,a>*/ {
    return handle(this.error)
  }
  and /*::<b>*/(other/*:Result<x,b>*/)/*:Result<x,b>*/ {
    return new Error(this.error)
  }
}
Error.prototype.type = "Error"


export const error = /*::<x, a>*/
  (error/*:x*/)/*:Result<x, a>*/ =>
  new Error(error)
```

### `stdout`

```
src/result.js.flow:6
  6:   , and <b> (result:Result<x, b>):Result<x, b>
                         ^^^^^^^^^^^^ intersection: type application of polymorphic type: class type: ResultInterface & object type. This type is incompatible with
 19:   and /*::<b>*/ (other/*:Result<x, b>*/)/*:Result<x, b>*/ {
                              ^^^^^^^^^^^^ union: type application of polymorphic type: type `Error` | type application of polymorphic type: type `Ok`. See: src/result.js:19
```

Now if modify `capture` implementation on `Ok` class in `result.js` as follows it type checks fine:

```diff
iff --git a/src/result.js b/src/result.js
index 81d4ea6..871aa2a 100644
--- a/src/result.js
+++ b/src/result.js
@@ -14,7 +14,7 @@ class Ok <x, a> {
     this.value = value
   }
   capture /*::<y>*/ (handle/*:(error:x) => Result<y, a>*/)/*:Result<y,a>*/ {
-    return this
+    return new Ok(this.value)
   }
   and /*::<b>*/ (other/*:Result<x, b>*/)/*:Result<x, b>*/ {
      return other
```

My guess is that flow fails to understand that  `Ok<x, a>` and `Ok<y, a>` are identical. In my real code I managed to workaround this by tricking flow to believe that I re-instantiate such types:

https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L28
https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L31
https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L37
https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L68
https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L71
https://github.com/Gozala/value-result/blob/0.0.1/src/result.js#L83

To summarise:

1. I would expect flow to handle such use cases without workarounds illustrated above. Meaning flow should be able to deduct that returning `this` is equivalent of reboxed version and there for not require reboxing.
2. Flow actually fails to type check blaming completely unrelated lines of code makes it insanely difficult to figure out the issue. I've spend more than a day dissecting code back and forth and still discovered what was going on by pure luck. I would expect flow to blame `capture` method for type incompatibility and a message like `Ok<x, a>` is not compatible with `Result<y, a>` or something along those lines.

**Comment by samwgoldman** [comment]
I would not expect Flow to handle this use case. The `.js.flow` stuff was not designed to support what you're doing here, which is to import `foo.js.flow` from `foo.js`, presumably to achieve a solution to #838. Gotta say, I really admire your persistence.

Your bug is unrelated to that, but I think it's a bit harder to see because of it. I spent more time than I should have looking into your code, and you're right that the issue is the inconsistency between `Ok<x, a>` and `Ok<y, a>`, but it's not true that they are identical. I'm not sure why you believe they are.

Glad you figured it out and your point about error messages is well taken, but I don't think this is a particularly useful issue to motivate improvements, since there's so much going on.

**Comment by Gozala** [comment]
> Your bug is unrelated to that, but I think it's a bit harder to see because of it. I spent more time than I should have looking into your code, and you're right that the issue is the inconsistency between Ok<x, a> and Ok<y, a>, but it's not true that they are identical. I'm not sure why you believe they are.

@samwgoldman I should have chosen different word. How about equivalent given the circumstances ? Given `Ok` class implementation `ok` is equivalent of `new Ok(ok.value)`. Also given that `Ok<x, a>` instance and `Ok<y, a>` instances are implementations of `Ok<a>` which satisfies `Result<y, a>` I would expect flow to cast `this` to `Result<y, a>` as method signature suggests vs requiring extra allocation.

It saddens me that real issue as this are just dismissed, as it takes a significant effort to track down bugs, simplify code to submit reproducible test case that at the end of the day is just discarded. Please also note that as you yourself pointed out this issue has nothing to do with `.js.flow` it's just simplifying test case itself takes significant time as it is and adding file merging & renaming  is just another extra mile to illustrate the issue, which I would be willing to take if I had a better support, but given current experience I'm discouraged to even submitting issues.
